flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/GlobalStatisticsAggregatorTracker.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.shuffle;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.Set;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataInputDeserializer;
+import org.apache.flink.core.memory.DataOutputSerializer;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * GlobalStatisticsAggregator is used by {@link DataStatisticsCoordinator} to collect {@link
+ * DataStatistics} from {@link DataStatisticsOperator} subtasks for specific checkpoint. It stores
+ * the merged {@link DataStatistics} result and uses set to keep a record of all reported subtasks.
+ */
+@Internal
+class GlobalStatisticsAggregatorTracker<D extends DataStatistics<D, S>, S> implements Serializable {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(GlobalStatisticsAggregatorTracker.class);
+  private static final double EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE = 90;
+  private final transient TypeSerializer<DataStatistics<D, S>> statisticsSerializer;
+  private final transient int parallelism;
+  private transient volatile GlobalStatisticsAggregator<D, S> inProgressAggregator;
+  private volatile GlobalStatisticsAggregator<D, S> lastCompletedAggregator;
+
+  GlobalStatisticsAggregatorTracker(
+      TypeSerializer<DataStatistics<D, S>> statisticsSerializer, int parallelism) {
+    this.statisticsSerializer = statisticsSerializer;
+    this.parallelism = parallelism;
+  }
+
+  boolean receiveDataStatisticEventAndCheckCompletion(
```

Comment:
I am wondering if this method signature is more clear? I found the class name like `AggregatorTracker` a little weird. With this signature, we can probably just call this class `Aggregator`. Then it only needs to maintain the in-progress aggregation. We can remove the other class (currently named as `Aggregator).

```
/**
 * Return globally aggregated statistics if received local statistics from all subtasks for the current checkpoint id or
 * if the in-progress aggregation is for a previous checkpoint and there are enough reported subtasks (90%).
 * Return null otherwise (as no completed aggregation).
 */
DataStatistics<D, S> merge(int subtask, DataStatisticsEvent<D, S> event);
```

Is this pseudo code logic a little clearer?
```
if (inprogressCheckpointId != null && inprogressCheckpointId > eventCheckpointId) {
  // ignore the event from older checkpoint
  return null;
}

DataStatistics completedStatistics = null;

if (inprogressCheckpointId != null && inprogressCheckpointId < eventCheckpointId) {
  if (over 90%) {
    completedStatistics = inprogressStatistics;
  } 
  inprogressCheckpointId = null;
  inprogressStatistics = null;
}

if (inprogressStatistics == null) {
  inprogressCheckpointId = eventCheckpointId;
  inprogressStatistics = new DataStatistics();
}

inprogressStatistics.merge(localStatistics);
if (all subtasks reported) {
  completedStatistics =  inprogressStatistics;
  inprogressCheckpointId = null;
  inprogressStatistics = null; 
}

return completedStatistics;
```